File: 04_GEMMA_GWAS/PrediXcanDosages2GEMMABIMBAM.R
#converts PrediXcan dosage format to GEMMA BIMBAM input format
"%&%" = function(a,b) paste(a,b,sep="")
library(data.table)

for(i in 1:22){
  geno <- read.table('zcat /home/angela/px_his_chol/Imputation/UMich/UMich_dosages/chr' %&% i %&% '.maf0.01.r20.8.dosage.txt.gz')
  geno$V1 <- NULL
  geno$V3 <- NULL
  geno$V6 <- NULL #column for MAF
  geno$V4 <- substr(geno$V4, 0, 1) #only one allele for minor and major
  geno$V5 <- substr(geno$V5, 0, 1)
  
  #for(j in 4:(length(geno))){ #if any value is above 2 or below 0, replace value w/ NA (which was apparently not the problem)
  #  geno[[j]] <- ifelse(geno[[j]] > 2 | geno[[j]] < 0, "NA", geno[[j]])
  #}

  fwrite(geno, '/home/angela/px_his_chol/GEMMA/BIMBAM/chr' %&% i %&% '.txt', row.names = F, col.names = F, quote = F, sep = " ")
  zip <- 'gzip /home/angela/px_his_chol/GEMMA/BIMBAM/chr' %&% i %&% '.txt'
  system(zip)
}

File: 04_GEMMA_GWAS/PrediXcanDosages2GEMMABIMBAM.py
#converts PrediXcan dosage format to GEMMA BIMBAM input format
import gzip
dosage_path = "/home/angela/px_his_chol/Imputation/UMich/UMich_dosages/"
BIMBAM_path = "/home/angela/px_his_chol/GEMMA/BIMBAM/"

for i in range(1,22):
    BIMBAM = gzip.open(BIMBAM_path + "chr" + str(i) + ".txt.gz", "wb")
    for line in gzip.open(dosage_path + "chr" + str(i) + ".maf0.01.r20.8.dosage.txt.gz", "rb"):
        arr = line.strip().split()
        (chr, rs, bp, A1, A2, MAF) = arr[0:6]
        if len(A1) < 2 and len(A2) < 2: #exclude multi-allelic
            dosages = arr[6:]
            dosages_str = '\t'.join(dosages)
            BIMBAM_format = (rs + "\t" + A1 + "\t" + A2 + "\t" + dosages_str + "\n")
            BIMBAM.write(BIMBAM_format)
    BIMBAM.close()
